\documentclass{PHlab-thesis}
\usepackage{amsmath}
\usepackage{graphicx}
\addbibresource{thesis.bib}

\newcommand*\Department中文{資訊工程學研究所}
\newcommand*\Department英文{Institute of Computer Science and Information Engineering}

\newcommand*\ThesisTitle中文{EAGLE-GPU：使用圖形處理單元加速計算基於 DNA 測序數據之推定基因組變異的統計分析}
\newcommand*\ThesisTitle英文{EAGLE-GPU： Using Graphics Processing Units to accelerate computation of the statistical support of putative genome variants based on DNA sequencing data}
%\newcommand*\ThesisNote中文{示例:其實徐翡曼是東京大學畢業的博士}% For real thesis omit, or use {初稿} etc.
%\newcommand*\ThesisNote英文{Just an example.  Fei-Man actually graduated from Tokyo Univ.}% For real thesis omit, or use {draft} etc.

\newcommand*\Student中文{盧宥霖}
\newcommand*\Student英文{You-Lin Lu}

\newcommand*\Advisor中文{賀保羅}
\newcommand*\Advisor英文{Paul Horton}

%% 果有共同指導老師可以用:
%% \newcommand*\CoAdvisorA中文{}
%% \newcommand*\CoAdvisorA英文{}
%% \newcommand*\CoAdvisorB中文{}
%% \newcommand*\CoAdvisorB英文{}


\newcommand*\YearMonth英文{July, 2022}
\newcommand*\YearMonth中文{１１１年７月}

\pagestyle{fancy}
\begin{document}


\newcommand*\Keywords英文{genomics, variant calling, GPU acceleration}
\newcommand*\Abstract英文{
Variant calling remains to play an important role in analyzing genome sequence data. A variety of studies have been conducted in the related field, proposing solutions to this specific task. The predecessor of this study, EAGLE: Explicit Alternative Genome Likelihood Evaluator, aimed to further improve the precision of the results of those, by assessing the likelihood of the called variants. It was shown that such technique assuredly enhance the accuracy. However, because of the computational complexity of its algorithm, the amount of time required to execute intensifies as the length of sequenced reads grow. Besides, such issue also acts as an obstacle when building more advanced applications on top of it. In this research, we would like to investigate if graphics processing units, or GPUs, could accelerate the process since the algorithm is highly parallel. Here, we propose EAGLE-GPU, a revision of the original method by rewriting the fundamental computing functions as GPU CUDA kernels. The experimental outcomes demonstrate that although the current usage of GPU parallelism might not be suitable for sequencing data with shorter reads, it is beneficial when the read lengths grow, opening up opportunities for the third generation sequencing. In addition, our work enables more complicated applications to establish on top with, due to the reduced time for the computational excessive operations. To conclude, we recommend applying GPU acceleration if encountering a highly parallel dataset with long reads, and that more specialized CUDA kernels could be designed base on our results for future complex problems.
}


\newcommand*\Keywords中文{基因組、變異位點偵測、平行運算}
\newcommand*\Abstract中文{
變異位點偵測在分析基因組序列數據中扮演相當重要的角色。在相關領域中已有各項研究陸續被提出，為這一特定任務提供可行的解決方案。本研究的前身 EAGLE：顯式替代基因組可能性評估器，旨在通過計算各個被偵測到潛在變異位點的可能性，以統計學方法來進一步提高結果的精確度。根據當時的研究數據，這種技術確實有效提高了準確性，相當適合用來進一步分析、處理各大變異位點偵測工具所產生的結果。然而，由於其演算法在計算上的時間複雜度，執行該程式所需的時間會隨著定序讀取長度的增加而大幅增長。此外，高時間複雜度的特性也讓我們在基於原始程式嘗試構建更高層級的應用程式時，遇到執行效率上的問題。由於我們認為該演算法是可以高度平行處理的，在這項研究中，我們想調查簡稱 GPU 的圖形處理單元，是否可以有效加速該過程。在這裡，我們提出了 EAGLE-GPU，通過將基本計算函式重新實作為 GPU CUDA 內核，透過平行處理底層機率計算的方式進行加速。為了未來的可擴充性，我們目前實作的 GPU 平行程式僅包含最底層的計算部分。依據實驗結果，目前的實作可能不適合對應較短的定序資料，若以次世代定序資料為主要研究對象，需要進一步設計更加專一的 CUDA 內核提高平行性。然而，當序列長度增加時，實驗數據顯示利用 GPU 做加速能大幅地降低所需的執行時間，對於序列長度超過一萬的第三代定序而言，無疑是一大幫助。此外，由於直接在基礎計算上減少了過多的時間消耗，我們提出的這份研究也可以對日後建立更高層級的複雜應用程序帶來幫助。綜合來說，根據這次的研究結果，我們建議在遇到高度並行的數據集時適當的應用 GPU 的輔助以加速程式的執行，並且在遭遇未來的複雜問題時，也可以根據我們的結果針對該問題的特性設計 CUDA 內核，快速有效的產生解決的方案。
}

\newcommand*\Acknowledgements{%
首先我想在這裡感謝我的指導教授賀保羅老師。對於大學期間沒有涉略生物相關基礎知識的我來說，透過老師的幫忙讓我了解基因序列分析常用的技巧與困難，也才衍生出進行這份研究的動機。在討論的過程中，也很感謝老師給予的指教，讓我在面對困難時更快速的找到解決問題的方向。本論文的完成同時還要感謝原論文 EAGLE 的主要作者之一 Tony Kuo，除了和我分享當時研究時的寶貴經驗，還提供了當初實驗的測試資料，使我們在本論文中能更客觀地針對結果進行比較與分析。最後，也要感謝這兩年來在實驗室一起奮鬥過的所有同學們，面對大家普遍陌生的生物資訊領域，彼此教學相長，透過不斷的交流討論更快速的習得各方面的專業知識。
}

\input{frontmatter}% 封面頁, 口委中英文簽名單, 誌謝, 中英文摘要, 論文目錄, 圖表目錄


%────────────────────  List of Symbols  ────────────────────
\renewcommand\nomgroup[1]{%
  \item[\bfseries
  \ifstrequal{#1}{A}{General}{%
  \ifstrequal{#1}{Z}{Gene/Protein Names}%
  }]}

% \nomenclature[A]{$\lg$}{Logarithm base 2}
% \nomenclature[A]{KL\ Divergence}{Kullback-Liebler Divergence}
% \nomenclature[Z]{Myc}{MYC proto-oncogene}
% \nomenclature[Z]{USF-1}{Upstream stimulatory factor 1}

\nomenclature[A]{NGS}{Next Generation Sequencing}
\nomenclature[A]{TGS}{Third Generation Sequencing}
\nomenclature[A]{DNA}{deoxyribonucleic acid}
\nomenclature[A]{CPU}{Central processing unit}
\nomenclature[A]{GPU}{Graphics processing unit}
\nomenclature[A]{EAGLE}{Explicit Alternative Genome Likelihood Evaluator, the predecessor of this research}
\nomenclature[A]{bp}{base pair}

\printnomenclature[5cm]

\newpage
\setcounter{page}{1}
\pagenumbering{arabic}


\chapter{Introduction}
\section{Background}
Next generation sequencing (NGS) has played a significant role in bioinformatics research for the past decades~\cite{behjati2013next,schuster2008next}. This technology enables sequencing the whole human genome sequence in a relatively short amount of time, allowing more advanced biological analysis and even medical applications~\cite{roukos2010next}. For example, numerous genetic and epigenetic mechanisms, such as DNA mutation and methylation, could be scrutinized if accurate genome sequencing data is available~\cite{moore2013dna}. Better knowledge of the variation in human genomes could have a great impact on clinical practice; notably the diagnosis of hereditary diseases and those related to genetic disorders~\cite{shashi2014utility,stenson2017human}, including specific subtypes of cancer~\cite{serrati2016next}.

Despite the breakthroughs made by utilizing NGS results, some computational challenges remain~\cite{muzzey2015understanding}.  DNA sequencing machines are unable to produce reads anywhere near as long as an entire human chromosome, so sequencing data consists of many (often millions) of shorter read segments originating from random fragments of the sequenced genome.  Fortunately a ``human reference genome'' is available, which usually allows computer programs to infer which fragment of the genome a read originated from.  This process however is complicated because: (1) sequence reads contain errors, (2) the human genome is repetitive, so fragments from different positions can look alike, and (3) individual genome differ from the human reference. As a consequence, the concept of read quality is introduced during the sequencing procedure, with its score indicating the probability of a base being misread.

Variant calling is the task of inferring differences between a sequenced individual's genome and the reference genome, which is a critical function for the analysis of NGS data. To finish this job, one would usually first collect two input data: the sequenced reads of an individual along with their read quality scores, and the reference genome. Then, sequence alignment is performed to map the sequenced reads to positions of the reference genome where they are most likely to be selected from. Finally, variants are called according to the alignment results, recognizing positions where the aligned reads conflict with the reference genome as potential variants.
Various methods and tools have been developed for this specific task, focusing on increasing the accuracy of the called variant results. For instance, the Genome Analysis Toolkit, often known as GATK~\cite{mckenna2010genome}, and SAMtools~\cite{li2009sequence}, are some of the most popular works for this certain task. Notwithstanding the efforts put in this problem, though, the results of those often turn out to have a rather low precision-recall score, while disagreements were encountered between outcomes from different variant calling pipeline~\cite{o2013low}. In order to justify whether the called variant is likely to happen, there are a number of methods proposed to systematically assess the results~\cite{hwang2015systematic,yu2013comparing}. Moreover, there are also several methods proposed, acting as an additional process to the variant calling pipeline, to revise the called variants and boost the accuracy to some extent. They were shown to improve the results of the called variants after being applied at certain stages in the pipeline. This includes the predecessor of this study, EAGLE: Explicit Alternative Genome Likelihood Evaluator~\cite{kuo2018eagle}, published in 2017 by Tony Kuo et al.  EAGLE is applied after the variant calling phrase, using a probabilitistic model of the sequencing process to compute a likelihood for each candidate variant.  In comparison to the originally candidate variants, they demonstrated that such post-processing can improve precision at an acceptable recall rate.

\section{Motivation & Research objective}
In a nutshell, this thesis explores the possibility of using Graphics Processing Units (GPU)s to accelerate EAGLE's computation.
The reason is that we noticed that the execution of EAGLE can take a considerable amount of time, even when using the multithreading option provided.  This execution time is still acceptable for the short reads (100--200 bp) provided by currently popular Illumina sequencing platforms.  However the EAGLE implementation was not
designed for upcoming third generation sequencing, where sequencing read lengths can be on the order of 10,000 bps in length.  Also EAGLE implements a method to consider insertion and deletion sequencing errors, but the authors
choose not to evaluate that, partially because they thought it would be too slow to be practical.
Thus we decided to investigate potential opportunities of acceleration.

Sequence alignment is a highly parallel task, and it was already shown in the original EAGLE research that multithreading could effectively speedup the whole process.  In order to further accelerate the program, adding more threads would be a straightforward approach.  But the total amount of threads that can be ran in parallel is constrained by the hardware, often capped around tens to twenties for modern mainstream CPUs.  This motivating us to search for other sources of acceleration, in particular the use of GPUs.

Originally designed for graphics, GPUs were first dedicated to graphics rendering pipeline.  They are responsible for tasks such as vertex and fragment shading, rasterization, etc.  The results are then mapped to the output devices, usually a computer monitor, where they are assembled to compute the value of each pixel, before the frame is presented to the user of the computer~\cite{foley1994introduction,owens2008gpu}.  This job, being very computational expensive, depends heavily on the throughput that can be produced by the processors.  Thus, the design of GPUs is mainly focused on the total amount of data throughput, aiming to provide as many frames per second as possible, in order to provide a smooth screen that results in a better user experience.  There are multiple ways to achieve better data throughput, applicable for different situations. When it comes to graphics rendering, which is a highly parallel task to compute values of RGB channels for each pixel presented to the screen, enabling more threads to execute the instructions concurrently is definitely a favorable direction. Hence, modern GPUs come with a lot more processing cores in comparison to CPUs. Although the computing power of a single GPU core is less than that of a single CPU core, with the total amount of cores outnumbering CPUs by a lot, many threads can be run at the same time due to having its own physical core, which is highly beneficial for data-parallel problems~\cite{navarro2014survey}.

Despite the dedication to graphics rendering when first developed, GPUs have gained their popularity towards general purpose computing. Equipped with enormous amount of physical cores, GPUs are also found suitable for single instruction, multiple data (SIMD) missions. With the size of data to be processed continuously thriving in the recent years, even simple instructions began to become computationally demanding. People started to look upon the utilization of GPUs' high-throughput architecture, seeking for the possibility to speed up time consuming jobs~\cite{owens2008gpu,nickolls2010gpu}. Although programming on GPUs were not straightforward due to quite a few limitations since they were used to tackle rendering problems in the past and programmability was never a main concern, NVIDIA, one of the main GPU designers, developed the well known CUDA architecture, acting as a integrated parallel computing platform with the application programming interface that comes collectively. This grants programmers an easier access to the GPU devices, encouraging them to explore the feasibility of parallel computation on GPUs and other sources of general purpose computation. In recent years, GPUs have been shown to greatly accelerate the process of a variety of tasks, including computer vision, deep learning, etc. Thus, we would like to investigate whether sequence alignment and variant calling, composed of several subtasks which are also highly data-parallel, can be accelerated with the aid of GPUs.

Here, we present EAGLE-GPU, a GPU accelerated version of EAGLE.  To better integrate with the existing toolchain, we adopted the EAGLE probabilistic model as is; but we did make several adjustments to reduce memory copy between device and host, making it a better fit to the device-host parallel computing architecture.  This makes EAGLE-GPU flexible for users from the previous program, allowing them to toggle between using the CPU solely or enabling GPU acceleration, according to the characteristics of the input data and the availability of GPU devices.  We conducted experiments using both newly simulated data and the identical data that the original EAGLE paper used; demonstrating that the likelihood evaluation computation can be signficantly speed-up for long reads.

\chapter{Related Works}
In this chapter, we would like to first review a few recent works related to GPU acceleration applied to bioinformatics in brief. With regards to this research being a follow up work of EAGLE, we would also take a closer look at the underlying probabilistic model behind the original method, which would later be used and modified to run on the GPU device.
\section{Bioinformatics with GPU}
\subsection{NVBIO}
The high throughput strength of GPU has already been applied to the field of bioinformatics. NVIDIA released their own GPU-based library, NVBIO, for sequence alignment problems~\cite{nvbio2015}. In virtue of the fine tuned functions in the library, users are supported to program in a relatively high level fashion, without needing to handle issues about physical GPU device access. This helps alleviate the frustration programmers, especially those who used to write code that runs exclusively on CPUs, often face with when encountering GPU related low level operations. Aside from libraries for programmers to integrate in their bioinformatics project, NVBIO accommodates several applications which were already built on top of the massively parallel architecture. Multiple well known sequence alignment algorithms were re-engineered in a high parallel manner, implemented with the previously developed library. For example, Bowtie2~\cite{langmead2012fast}, a gapped read alignment tool, is recreated as nvBowtie in the package.
\subsection{Others works}
Numerous other applications in bioinformatics were also shown to be advantageous with GPUs. Yung et. al presented GBOOST~\cite{yung2011gboost} for gene-gene interaction, while an influential sequence analysis tool, BLAST~\cite{altschul1990basic}, were implemented with parallel computing based on GPU and achieved around 2 to 4 times speedup~\cite{vouzis2011gpu}. Several works, on the other hand, commit to providing GPU libraries with even better performance. GASAL2~\cite{ahmed2019gasal2}, for instance, is another accelerated library for sequence alignment, reported to outperform NVBIO. 

\section{EAGLE}
Critical concepts of the previous work EAGLE would be recapitulated in this section. First of all, three files would be provided as input: the candidate variants in a vcf format, the reference genome, and the aligned sequencing reads in a bam file format. After processing the data given, the candidate variants are then evaluated by their likelihood. With the given data, we could obtain the quality of candidate variants by examining whether the reads better support the hypothesis which they were sequenced from the reference genome, or rather being more likely to be sequenced from an alternative sequence consisting of the candidate variants, indicating a higher probability that the candidate variants are actually present. On the other hands, if the reads better suit the other hypothesis that they are not related to the candidate variants and were more possibly sequenced from the reference genome, we could then lower our confidence level on the candidate variants. Therefore, we would have to calculate the probability $P[r|G]$ regularly over all sequencing reads $r$ in a read set covering the interval with interest, while $G$ denotes the hypothetical genome sequence, either one of the reference genome $G_{ref}$ or the alternative genome, $G_{alt}$. For each iteration calculating $P[r|G]$, we can derive $P[r|G]$ as the summation of the probability having $r$ being sequenced from $g$, a certain genome segment in $G$ with the same length $\ell_r$ as the read $r$, over all possible ${g\in G}$. Such equation can be written as follows:
\begin{equation}
P[r|G] = \sum_{g\in G} P[g|G] P[r|g]
\end{equation}
where $P[g|G]$ is the probability that genome sequence segment $g$ is sequenced, with respect to that some segment from $G$ is sequenced. In the original work, two assumptions were made here. Firstly, uniform priors were assumed on genome segments $g$ to be in any particular length. Secondly, to acquire the probability that the genome segment were having the same length as the read, available for further computation, the length distribution of $g$ is assumed to follow the general prior defined as $\mathcal{L}(\ell)$. Which is, in mathematical equation,
%%
\begin{equation}
P[g|G] = \mathcal{L}(\ell_g) \frac{1}{n_{\ell_g}}
\end{equation}
having the symbol $n_{\ell_g}$ representing the number of genome segments of length $\ell_g$.
Now there is $P[r|g]$ remaining in the previous equation. Regarding that it is difficult to compute directly, Bayes' Law could be applied, leading to
\begin{equation}
P[r|g] = \frac{P[r]}{P[g]} P[g|r]
\end{equation}
Here, $P[g]$ denotes the prior probability of sequence $g$ given that $g$ was read by a sequencer, considering possible errors when sequencing. However, an assumption that no indel errors happened was made back then, assuring that the length of the sequence did not change during the process. The prior probability distribution of the length of reads, denoted as $\mathcal{L}(\ell)$, is resolved during the experiment. For DNA sequencing, there are 4 potential options in the set \texttt{\{a,c,g,t\}} for each base, with $\ell_g$ bases to be considered, which is,
\begin{equation}
P[g] =  \frac{\mathcal{L}(\ell_g)}{4^{\ell_g}}
\end{equation}
Here, since $P[r]$ is going to be canceled later on, it will be left as is. Aware of sequencing errors, the computation of $P[g|r]$ is performed by calculating the product of the probabilities for all bases, considering the base call quality score. Thus,
\begin{equation}
P[g|r] = \prod_{i=1}^{\ell_r} P[g_i|r_i].
\end{equation}
where $g_i$ denotes the $i$th base of $g$, while $r_i$ represents the probability vector for the $i$th position of $r$, with respect to the read quality score over \texttt{\{a,c,g,t\}}. The above equations can thus further derive
\begin{equation*}
\begin{split}
P[r|G] =  \sum_{g\in G} P[g|G] P[r|g] &= \sum_{g\in G} \left( \mathcal{L}(\ell_g) \frac{1}{n_{\ell_g}} \right) \frac{P[r]4^{\ell_g}}{\mathcal{L}(\ell_g)} \prod_{i=1}^{\ell_g} P[g_i|r_i]  \\
                                      &= \sum_{g\in G} \frac{P[r]4^{\ell_g}}{n_{\ell_g}} \prod_{i=1}^{\ell_g} P[g_i|r_i]  \\
                                      &= \sum_{g\in G} \frac{P[r]}{n_{\ell_g}} \prod_{i=1}^{\ell_g} 4P[g_i|r_i]  \\
                                      &\approx \frac{P[r]}{n} \sum_{g\in G} \prod_{i=1}^{\ell_g} 4P[g_i|r_i]  \\
\end{split}
\end{equation*}

It is obvious that, the above equation requires computing the summation of products over the length $\ell_g$. This is where the iterative loops of data parallel operations were executed. Focusing on reducing the time required to complete the above task, we will describe the method adapted in the following chapter.

\chapter{Method}
\section{Method Overview}
In the original EAGLE publication, whether a hypothesis fits the sequenced read data is determined by its likelihood, often measured by its ratio against the default hypothesis: the read being sequenced from the reference genome. In order to compute the likelihood for all possible read alignments, the probabilities of all read and genome segment pair are calculated and then further summed up in logarithm. Although this is undoubtedly a straightforward solution, numerous time consuming operations were done in loop. Here, we unroll the nested loops and take advantage of the high parallel graphics processing units. 

% \includegraphics[width=6cm]{figures/overview.png}

\begin{figure}
	\centering
	\includegraphics[scale=0.3]{figures/overview.png}
	\caption{Method overview}
	\label{fig:overview} % \ref{this label}
\end{figure}

As shown in Figure \ref{fig:overview}, our goal is to parallelize the parts circled, with the attempt to reduce the total amount of time required to execute the program. We will go through the details in the following section.

\section{Proposed Scheme}
\subsection{CUDA programming model}
Before all else, we should first cautiously design our workflow, trying to acclimate to the CUDA programming model. There exists several essential points to be aware of, which could affect the performance significantly because of the hardware implementations of GPUs. For modern NVIDIA GPUs that support CUDA, often comes with numerous streaming processors(SP), which are the cores on the GPU. Each of them is responsible for executing a thread in our program. They are grouped to form streaming multiprocessors(SM), the basic unit programmers control to execute our parallelized functions, also known as kernels in CUDA terminology. Streaming multiprocessors have their own set of streaming processors, along with their own hardware units such as registers, cache, shared memory, and warp scheduler that schedules warps, the basic unit when it comes to runtime execution. Each warp contains 32 threads, while they share one program counter. Thus, when threads in the same warp executes instructions conditionally, possibly through statements such as if-else, will slow down the process since those threads not executing will be temporarily deactivated, waiting for others to finish. Such problem is referred to as warp divergence, and could result in a performance loss due to the serialized manner. Another issue that could worsen performance is memory copying. Since kernels launched on GPU, or device in CUDA terminology, cannot directly access the host memory, those data to be computed should be copied to the device beforehand. However, copying data between host and device is quite time consuming, being likely to take more time than the actual computing operations if not designed carefully. Hence, these are our main concerns during development, while seeking to keep the kernels generalized, ready for further integration. To better demonstrate how the task is designed in a parallel fashion, we briefly visualized our implementation fitting the CUDA programming host-device model in Figure \ref{fig:prop}. 
\begin{figure}
	\centering
	\includegraphics[scale=0.8]{figures/proposed.png}
	\caption{Proposed workflow in host-device scheme}
	\label{fig:prop} % \ref{this label}
\end{figure}
% To better demonstrate how the task is designed in a parallel fashion, the pseudocode of the implementation, without parallel computing, is shown below. 
% \begin{algorithm}
% 	\caption{An algorithm with caption}\label{alg:cap}
% 	\begin{algorithmic}
% 	% \Require $n \geq 0$
% 	% \Ensure $y = x^n$
% 	\For{each $g\in G$}
% 		\For{$i=1$ to $\ellr$}
% 			\State $//$ Your source here;
% 		\EndFor
%     \EndFor
% 	\end{algorithmic}
% \end{algorithm}

We will then describe how we built up the parallelized implementation step by step in the following sections.

\subsection{Computing the likelihood with given genome segment}
First of all, for any given pair of genome segment $g$ and sequence read $r$, the likelihood of the read being sequenced from $g$ is computed by product of the likelihood of all base pairs. Since the likelihood of each base pair match/mismatch is independent to that of other base pairs, they can be computed concurrently. This provides us the opportunity to accelerate this process by computing them in parallel, where the maximum number of operations that can be done in parallel is bounded by the length of the read $r$. Considering NGS read data, which is often around 100 to 150 base pairs in terms of read length, we can easily load the entire read onto the GPU and allocate a thread for each base, since the size is rather small in comparison to the memory capacity and the core counts of modern GPUs. A schematic diagram for such process is shown in Figure \ref{fig:prg}.
\begin{figure}
	\centering
	\includegraphics[scale=0.3]{figures/method1.png}
	\caption{Parallel computation for $P[r|g]$}
	\label{fig:prg} % \ref{this label}
\end{figure}


\subsection{Computing the likelihood with given reference genome}
Next, we would like to further extend the degree of parallelism, seeking to utilize the performance provided by the GPU. Here, we take a closer look at the intermediate result, $P[r|G]$. In most cases, knowing the exact genome segment $g$ where read sequence $r$ is sequenced from in advance is unrealistic. Thus, in the original method, for all possible genome segment $g$ sampled from the hypothesis genome $G$, with the exact length with the given read $r$, were considered. $P[r|g]$ are calculated iteratively, with the summation of their results being the final score of the given read. Since the total amount of segments to be considered is related to the length of read sequence ($2{\ell_r}$), the execution time required grows in polynomial time with the sequence length. Apparently, the results of $P[r|g]$ for different genome segment $g$ is also independent, opening an opportunity for parallelism. In addition to the entire read $r$, we also loaded the selected genome $G$ to the GPU, allocating a block of threads for each possible genome segment $g$. Within each block, we execute the same instructions as described in the previous section, calculating the probabilities for each site in their corresponding thread, as shown in Figure \ref{fig:prG}. From the perspective of each thread, the probability they need to compute is illustrated in Figure \ref{fig:tv}.
\begin{figure}
	\centering
	\includegraphics[scale=0.3]{figures/prG.png}
	\caption{Multiple CUDA blocks when parallel computing $P[r|G]$}
	\label{fig:prG} % \ref{this label}
\end{figure}
\begin{figure}
	\centering
	\includegraphics[scale=0.3]{figures/threadview.png}
	\caption{Tasks for each thread when parallel computing $P[r|G]$}
	\label{fig:tv} % \ref{this label}
\end{figure}

Through the above method, we have already reduced the amount of time required to execute the program. This also allows us to revisit a term that was previously ignored: insertion/deletion errors in read sequences. Since indel errors not only modifies the sequence but also introduce shifts, it would increase the time complexity, proportional to the length of reads. With the previous implementation, where the operations were sequentially executed, this would sure intensify the computational cost and lead to longer execution time. Mainly considering generation sequencing data, which has a rather low probability of indel errors in the sequencing phase, an assumption that there was no indel sequencing errors was made, in order to avoid the problem described above. However, with the aid of GPUs, such operations can be done concurrently, reducing the amount of time required to compute the final result. In GPU-EAGLE implementation, a threshold value is input by the user, indicating the maximum number of indel errors to be taken into account. After all, the probability of such errors is still fairly low, making it pointless to compute all of the combinations of shifted reads and indels. Knowing the threshold beforehand, all of the possible reads having less indel errors than it is then listed out and assigned to the GPU. Similar to the base case where no indel errors was considered, we would need a block of threads for each pair of read r and genome segment g. With regard to the same read having multiple variations due to the introduced indel errors, the GPU would launch a two dimensional block instead of the original one dimensional block, with the extra dimension representing the index of read in the list of possible reads with indel. To allow such indeled read list to be long exceeding the block size launched, either because of hardware constraints or user settings, the list would be partitioned to batches with the same size as the y dimensional size of the two dimensional block. At last, the results of each batch would be added up to acquire the final result.

\chapter{Results}
\section{Input dataset}
To better compare the results with and without GPU acceleration, an experiment was conducted using the reconstructed diploid sequence of human genome chromosome 22 from hg19 reference sequence, which is one of the dataset that were used in the original study. Intend to investigate possible speedup when facing third generation sequencing read data, simulated sequencing reads with 10,000 base pairs were generated by random sampling from the reference genome. Reads with different lengths were also generated as a reference and an indicator for possible future generation sequencers.

\section{Benchmark platforms}
EAGLE-GPU is compiled with CUDA version 10.1. The experiments were conducted on a machine with a single AMD Ryzen™ 9 5950X desktop processor, consisting of 16 hyper-threaded physical cores running at their base clock 3.4 gigahertz, and 128 GB of random access memory. The GPU installed on the machine is a single NVIDIA GeForce RTX™ 3090 graphics card.

\section{Identical dataset: The NS12911 variants}
First of all, we used the exact same variant dataset that was mentioned in the original EAGLE paper to inspect the performance difference between the sequential and the parallelized implementations. In search of the best grid size and block size settings for this task, we conducted a first experiment disabling the CPU's multithreading setting, measuring the time taken for all combinations of 1, 32, 64, 128, 256, threads and blocks respectively. The measurements are shown in Table \ref{table:1}. As we can see in Figure \ref{fig:gpuconfig} where we visualized the results, the best configuration would be using 32 blocks per grid, with each block running 32 threads. This results in the program taking 0.342212 hours to finish the task. Later on, attempting to compare the results with and without GPU acceleration, the experiment is then conducted by adopting a single threaded CPU setting, with and without GPU parallel computing under the best kernel launch configuration, gradually moving to a multithreaded setting. The results are shown in Figure \ref{fig:multithread}. For this dataset, the execution time both reduced with the amount of threads increasing, until 8 threads were used. However, we can see that with the current implementation of GPU computing, the execution time is not effectively reduced, but is somehow worsened instead.
\begin{table}[h!]
	\centering
	\begin{tabular}{l*{5}{c}}
		Block size              & 1 & 32 & 64 & 128 & 256\\
		\hline
		1 block & 0.363213 & 0.351147 & 0.350840 & 0.349051 & 0.346477  \\
		32 blocks            & 0.364184 & 0.342212 & 0.345006 & 0.344206 & 0.344719  \\
		64 blocks          & 0.353136 & 0.353382 & 0.347007 & 0.349023 & 0.345956  \\
		128 blocks     & 0.354095 & 0.353936 & 0.352839 & 0.349773 & 0.349343  \\
		256 blocks     & 0.351260 & 0.355450 & 0.359848 & 0.361534 & 0.357256  \\
	\end{tabular}
	\caption{Time taken (in hours) for different kernel launch configuration}
	\label{table:1}
\end{table}

\begin{figure}
	\centering
	\includegraphics[scale=0.3]{figures/gpu_config.png}
	\caption{Time taken for different kernel launch configuration}
	\label{fig:gpuconfig} 
\end{figure}
\begin{figure}
	\centering
	\includegraphics[scale=0.3]{figures/multithread.png}
	\caption{Time taken for different CPU threading settings}
	\label{fig:multithread} 
\end{figure}

\section{Simulated reads with different lengths}
With the length of the sequenced reads growing considering third generation sequencing, we performed another experiment, with the intention to figure out the relations between parallelism and sequence length. We did not only test with sequences with around 10000 bp which is the most common length for Nanopore sequencing results, but also other varying lengths, preparing our tool in advance for possible future sequencing technologies which might produce reads in varying lengths. Simulated reads were generated, with different lengths, by randomly sampling from the reference genome. Concentrating on the performance difference regarding the presence of GPU acceleration only, we performed this experiment in a single CPU threading setting, with 256 blocks x 256 threads kernel launch configuration. Clearly seen in Figure \ref{fig:readlen}, the execution time without GPU acceleration dramatically increases with the read length, while that of GPU accelerated version growing rather mild, outperforming the other after reads expand over 1000 base pairs.
\begin{figure}
	\centering
	\includegraphics[scale=0.3]{figures/readlen.png}
	\caption{Execution time for different input read lengths}
	\label{fig:readlen} 
\end{figure}

\section{Considering read indel errors}
Attempting to find out if the consideration of indel errors is feasible, although computationally expensive, with the assistance of GPU parallel computing, we reused the NS12911 dataset, but without any assumption about the presence of read indel errors. Instead, a threshold is issued, while all of the reads with indel errors length within the threshold is considered. As mentioned in the previous study, doing so with CPU only is extremely time-consuming, thus we had to set a limitation for the threshold, not exceeding 1, in order to regulate the total execution time to a reasonable extent. As we can see in Figure \ref{fig:readindel}, a positive result of GPU speedup is achieved even with one possible read indel error, not to mention computing even more combinations of indels for a larger threshold.
\begin{figure}
	\centering
	\includegraphics[scale=0.3]{figures/read_indel.png}
	\caption{Time taken for different indel threshold settings}
	\label{fig:readindel} 
\end{figure}

\chapter{Discussion}
EAGLE-GPU is a GPU accelerated implementation of the previously published method, EAGLE, while maintaining similar data processing workflow for better integration with the previous tool. Being generalized by re-engineering the fundamental functions, not packing the entire algorithm into a hard coded GPU kernel sure helps with flexibility, preparing the tool for further applications. However, performance were sacrificed to a certain extent by doing so. In the following sections, we would like to discuss our interpretation of the experimental results, along with possible approaches to further increase the speedup.

\section{Case study: The original dataset}
As mentioned in the previous chapter, we can easily see that applying GPU computation does not benefit the results when the dataset tested in the previous research was used. Through observations of the profiling results, it is clear that the computational time, excluding the time for data copying between host and device, is longer for GPUs, not to mention data copying adds up to the total execution time if GPU is used. The main reason for this is that, although equipped with a lot more processors, the computational power of a single GPU core is weaker than that of a CPU core. As a consequence, if the extent of parallelism was not high enough, distributing tasks to the GPU could possibly become a drawback when it comes to execution time. To better illustrate such problem, a rather easy experiment was made, which is the addition of two vectors, with the execution time measured for different vector sizes. The results are shown in Figure \ref{fig:cpugpuvec}.
\begin{figure}
	\centering
	\includegraphics[scale=0.3]{figures/cpugpu_vector.png}
	\caption{Comparing the performance between CPU and GPU}
	\label{fig:cpugpuvec} % \ref{this label}
\end{figure}
As we can see in Figure \ref{fig:cpugpuvec}, applying GPU acceleration is not favoring until the size of vector exceeds 800000. Now we further inspect our dataset, where all of the sequencing reads have a length of 100 bp. When we calculate $P[r|G]$, a total of 100 x 100 = 10000 operations can be executed concurrently, each computing the probability at a certain cite. Although the time taken between addition and querying the probability of a DNA base might differ, but the considerable difference between the two numbers leads us to infer that the reads in this specific dataset is too short for it to be profitable when it comes to GPU computation. Such supposition could be confirmed with the second experiment, which we would further discuss in the next section. As a result, we suggest to not apply GPU acceleration when handling datasets with shorter reads. Since the instructions performed recurrently are rather little, the CPU can effectively finish the job solely, without needing any help from the GPU. The only condition we could think of that GPU acceleration would be profitable with short reads is when numerous reads maps to a single interval of interest. Under this circumstance, implementing a specialized kernel, computing $P[R|G]$, where $R$ denotes the set of reads, could be a practical approach. However, the main objective for this research is to build a rather general solution, for future applications to build on, thus this will be left as a possible future work, which can be conducted in a similar fashion if such type of data is encountered.

\section{Case study: Simulated TGS read data}
By conducting this experiment, we could not only validate the presumption described in the previous section, but also acquire an estimation of the possible speedup when applying EAGLE-GPU to third generation sequencing data. Here, we can easily see that the speedup performance gradually increases as the size of sequenced reads grow, surpassing it of the CPU when the read length passes 1000 bp. This sure supports our presumption, concluding that GPU acceleration is a favorable option when it comes to third generation sequencing, or even further sequencing technologies with sequencing read lengths beyond 1000 bp.

\section{Case study: Read indel errors}
The execution time dramatically increases when read indels are taken into consideration, justifying the attempt to ignore such errors in the previous work. With the aid of GPUs, such difficulty could be tackled, to a certain extent. As long as it fits into the device memory, we can load all possible reads, including different combinations of indels, onto the GPU. They can be computed concurrently, greatly reducing the time required for such operations, despite being computationally expensive. With the threshold adjustable, GPU acceleration could come in as a relief if the sequencing technology is reported to have a noticeable read indel issue.

\section{Future Work}
Trying to investigate various possibilities of application for GPU to improve the performance, this paper provides the implementations of the general usage functions in CUDA. Unlike programs executed particularly on CPU, the design of GPU kernels often require extra attention, including memory copying schema and the detail implementation of algorithms to avoid warp divergence. If we purely consider datasets with shorter sequencing reads, the extent of parallelism could be further increased by computing different read sets concurrently. With a specific kernel for each case, warp divergence could be eliminated. Theoretically, this could benefit cases when the dataset consist of large amount of read sets, enormously reducing the amount of time to evaluate the likelihood by computing the results for each set all at once. Another possible optimization would be lessening the memory copying cost. Despite it not being the bottleneck of the program, some advanced techniques could be applied to lower the amount of data copied between the host CPU and the device even more. For example, if we are targeting only DNA and RNA sequences but not proteins, we could easily represent the bases, A, C, G, T(U for RNA), N, in 3 bits. Thus, instead of copying them as characters, they could be encoded first, enabling us to pack 2 bases into one unsigned integer. This is a common trick among sequencing related GPU applications, also adopted by other tools such as GASAL2~\cite{ahmed2019gasal2} and NVBIO~\cite{nvbio2015}.

\chapter{Conclusion}
In this paper, we propose and implemented a GPU accelerated version of EAGLE, a previously proposed method for alternative genome likelihood evaluation. GPU-EAGLE utilizes the massively parallel architecture of modern GPUs, enabling instructions to be executed concurrently on multiple different data. This prepares the original method for the continuous growth when it comes to the size of sequencing data. Besides, we also investigated the feasibility of taking sequencing read indel errors into consideration, which is computational expensive and used to be ignored, after increasing the program throughput by applying parallel computation. We provide experimental results showing that GPU-EAGLE outperforms the CPU version when the data size is large enough, and conclude that designing a specialized GPU kernel could always be an option for forthcoming applications with excessive computational operations.

\newpage
\AddToContents{References}
\printbibliography


\end{document}
